distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso

```diff
@@ -0,0 +1,586 @@
+        handle_java_exception <| handle_number_format_exception <|
+            case mc of
+                _ : Math_Context -> Decimal.Value <| handle_precision_loss s <| Decimal_Utils.fromString s mc.math_context
+                _ : Nothing -> Decimal.Value (Decimal_Utils.fromString s)
+
+    ## GROUP conversions
+       Construct a `Decimal` from an `Integer`.
+
+       Arguments:
+       - i: The `Integer` to construct a `Decimal` from.
+       - mc: The `Math_Context` to use to specify precision and `Rounding_Mode`.
+         If a `Math_Context` is used, there is a possibility of a loss of
+         precision.
+
+       ! Error Conditions
+
+         - If the construction of the Decimal results in a loss of precision, a
+           `Loss_Of_Numeric_Precision` warning is attached. This can only happen
+           if a `Math_Context` value is explicitly passed.
+
+        ^ Example
+          Create a `Decimal` from an integer.
+
+              d = Decimal.from_integer 12
+    from_integer : Integer -> Math_Context -> Decimal ! Arithmetic_Error
+    from_integer (i : Integer) (mc : Math_Context | Nothing)=Nothing -> Decimal ! Arithmetic_Error =
+        handle_java_exception <|
+            case mc of
+                _ : Math_Context -> Decimal.Value <| handle_precision_loss i <| Decimal_Utils.fromInteger i mc.math_context
+                _ : Nothing -> Decimal.Value (Decimal_Utils.fromInteger i)
+
+    ## GROUP conversions
+       Construct a `Decimal` from a `Float`.
+
+       Arguments:
+       - f: The `Float` to construct a `Decimal` from.
+       - mc: The `Math_Context` to use to specify precision and `Rounding_Mode`.
+         If a `Math_Context` is used, there is a possibility of a loss of
+         precision.
+
+       ? Precision Loss
+
+         A `Decimal` can represent any `Float` precisely. However, some `Float`
+         literals, such as `0.1`, cannot be represented exactly to infinite
+         precision by a `Float`. For this reason, constructing a `Decimal` from
+         a `Float` always attaches a `Loss_Of_Numeric_Precision` warning to the
+         result.
+
+       ! Error Conditions
+
+         - A `Loss_Of_Numeric_Precision` warning is always attached when
+           converting to `Decimal` from `Float`.
+         - If `f` is NaN or +/-Inf, an Illegal_Argument error is thrown.
+
+        ^ Example
+          Create a `Decimal` from a float.
+
+              d = Decimal.from_integer 12.345
+    from_float : Float -> Math_Context -> Decimal ! Arithmetic_Error ! Illegal_Argument
+    from_float (f : Float) (mc : Math_Context | Nothing)=Nothing -> Decimal ! Arithmetic_Error ! Illegal_Argument =
+        is_exceptional = f.is_nan || f.is_infinite
+        if is_exceptional then Error.throw (Illegal_Argument.Error "Cannot convert "+f.to_text+" to a Decimal") else
+            # We do not use `handle_precision_loss` here because we always attach a `Loss_Of_Numeric_Precision` warning in the Float case.
```

Comment:
related: https://github.com/enso-org/enso/pull/9272/files#r1521712933

IMHO if we changed to more unified handling, this comment might not have been necessary.
We also reduce the risk of forgetting to manually attach the warning in some other usage of `fromFloat`.

Reply:
Done